File: plc4j/transports/tcp/src/main/java/org/apache/plc4x/java/transport/tcp/TcpTransportInstance.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.plc4x.java.transport.tcp;

import org.apache.plc4x.java.spi.transports.api.AsyncTransportInstance;
import org.apache.plc4x.java.spi.transports.api.BaseTransportInstance;
import org.apache.plc4x.java.spi.transports.api.RingBuffer;
import org.apache.plc4x.java.spi.transports.api.exceptions.TransportException;
import org.apache.plc4x.java.spi.utils.StaticHelper;
import org.apache.plc4x.java.transport.tcp.config.TcpTransportConfiguration;
import org.apache.plc4x.java.utils.auditlog.api.AuditLog;
import org.apache.plc4x.java.utils.auditlog.api.AuditLogEventType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.AsynchronousCloseException;
import java.nio.channels.SocketChannel;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.LockSupport;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;

/**
 * Java 21+ TCP transport using one virtual thread per connection doing blocking
 * {@link SocketChannel} reads/writes. On Java 21 a virtual thread blocked in a blocking-mode
 * channel read/write parks and releases its carrier (the JDK registers the fd with the NIO
 * poller), so there is no NIO {@link java.nio.channels.Selector}, no readiness-event loop, and
 * no busy-wait. The public surface and the {@link AsyncTransportInstance} callback contract are
 * unchanged: the read loop fills the {@link RingBuffer} (under {@code readLock}) and invokes the
 * registered data listener exactly as the previous selector loop did.
 */
public class TcpTransportInstance extends BaseTransportInstance<TcpTransportConfiguration> implements AsyncTransportInstance<TcpTransportConfiguration> {

    private static final Logger LOGGER = LoggerFactory.getLogger(TcpTransportInstance.class);
    private static final int DEFAULT_BUFFER_SIZE = 81920;
    private static final byte[] EMPTY_BYTES = new byte[0];

    private final SocketChannel socketChannel;
    private final RingBuffer ringBuffer;
    private final ByteBuffer readBuffer;  // Reused per-connection direct buffer for channel reads (confined to the read thread)
    private final Lock readLock = new ReentrantLock();
    private final Lock writeLock = new ReentrantLock();
    private final AtomicBoolean open = new AtomicBoolean(true);

    // Async support
    private volatile Runnable dataListener;
    private volatile Consumer<Throwable> disconnectListener;
    private final Thread readThread;

    public TcpTransportInstance(InetSocketAddress remoteAddress, TcpTransportConfiguration configuration, AuditLog auditLog) throws TransportException {
        super(configuration, auditLog);
        LOGGER.debug("TcpTransportInstance");
        this.ringBuffer = new RingBuffer(configuration.receiveBufferSize);
        this.readBuffer = ByteBuffer.allocateDirect(DEFAULT_BUFFER_SIZE);  // Reused direct buffer for channel reads

        // Opened into a local first, only promoted to the final `socketChannel` field once every
        // setup step (bind, socket options, connect) has succeeded.
        SocketChannel channel = null;
        try {
            // Open socket channel
            channel = SocketChannel.open();

            // Bind to a local address if specified
            if (configuration.localAddress != null && !configuration.localAddress.isEmpty()) {
                SocketAddress localAddr = new InetSocketAddress(configuration.localAddress, configuration.localPort);
                channel.bind(localAddr);
                LOGGER.debug("Bound to local address {}:{}", configuration.localAddress, configuration.localPort);
            }

            // Configure socket options before connecting
            channel.socket().setTcpNoDelay(configuration.tcpNoDelay);
            channel.socket().setKeepAlive(configuration.keepAlive);

            if (configuration.sendBufferSize > 0) {
                channel.socket().setSendBufferSize(configuration.sendBufferSize);
            }
            if (configuration.receiveBufferSize > 0) {
                channel.socket().setReceiveBufferSize(configuration.receiveBufferSize);
            }
            // Note: configuration.readTimeout is intentionally NOT mapped to Socket.setSoTimeout here.
            // SO_TIMEOUT has no effect on blocking SocketChannel reads, so the previous call was a
            // silent no-op. Read timeouts are enforced at the protocol/driver layer (e.g. the S7 driver
            // bounds responses via CompletableFuture timeouts), not by the transport.

            // Connect with timeout
            channel.socket().connect(remoteAddress, configuration.connectTimeout);

            // Blocking mode: on Java 21 a virtual thread blocked in read()/write() parks and
            // releases its carrier, so no selector is needed.
            channel.configureBlocking(true);

            this.socketChannel = channel;

            LOGGER.info("Connected to {}:{} with async support", remoteAddress.getHostName(), remoteAddress.getPort());

            auditLog.write(AuditLogEventType.CONNECT, String.format(
                "Connected to: %s:%d with local address: %s:%d",
                remoteAddress.getHostName(), remoteAddress.getPort(),
                getLocalAddress().getHostName(), getLocalAddress().getPort()));

            // Start the per-connection read loop on a virtual thread (Java 21+) LAST, so an
            // unchecked throw from the logging/audit above cannot leak an already-running thread
            // (the catch only handles IOException and does not stop the read loop).
            this.readThread = Thread.ofVirtual()
                .name("TCP-Read-" + remoteAddress.getHostName() + ":" + remoteAddress.getPort())
                .start(this::runReadLoop);
        } catch (IOException e) {
            String errorMsg = String.format("Failed to connect to %s:%d - %s",
                remoteAddress.getHostName(), remoteAddress.getPort(), e.getMessage());
            LOGGER.error(errorMsg, e);
            // errorMsg already embeds e.getMessage(); a single audit event avoids a duplicate.
            auditLog.write(AuditLogEventType.ERROR, "Error in constructor: " + errorMsg);
            if (channel != null) {
                try {
                    channel.close();
                } catch (IOException closeException) {
                    e.addSuppressed(closeException);
                }
            }
            throw new TransportException(errorMsg, e);
        }
    }

    public InetSocketAddress getRemoteAddress() {
        return (InetSocketAddress) socketChannel.socket().getRemoteSocketAddress();
    }

    public InetSocketAddress getLocalAddress() {
        return (InetSocketAddress) socketChannel.socket().getLocalSocketAddress();
    }

    @Override
    public boolean isOpen() {
        return open.get() && socketChannel.isConnected();
    }

    @Override
    public int getNumBytesAvailable() throws TransportException {
        readLock.lock();
        try {
            if (!isOpen()) {
                return 0;
            }

            return ringBuffer.availableForReading();
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public byte[] peekReadableBytes(int numBytes) throws TransportException {
        if (numBytes <= 0) {
            return EMPTY_BYTES;
        }

        readLock.lock();
        try {
            ensureOpen();

            if (ringBuffer.availableForReading() < numBytes) {
                throw new TransportException(
                    String.format("Requested %d bytes but only %d available", numBytes, ringBuffer.availableForReading())
                );
            }

            // Peek without consuming
            return ringBuffer.peek(numBytes);
        } catch (TransportException e) {
            getAuditLog().write(AuditLogEventType.ERROR, "Error in peekReadableBytes: " + e.getMessage());
            throw e;
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public byte[] read(int numBytes) throws TransportException {
        if (numBytes <= 0) {
            return EMPTY_BYTES;
        }

        readLock.lock();
        try {
            ensureOpen();

            if (ringBuffer.availableForReading() < numBytes) {
                throw new TransportException(
                    String.format("Requested %d bytes but only %d available", numBytes, ringBuffer.availableForReading())
                );
            }

            // Read and consume bytes
            byte[] bytes = ringBuffer.read(numBytes);

            // Log the bytes to the audit log
            if (getAuditLog().isEnabled()) {
                getAuditLog().write(AuditLogEventType.INCOMING_BYTES, StaticHelper.ENCODE_HEX(bytes));
            }

            return bytes;
        } catch (TransportException e) {
            getAuditLog().write(AuditLogEventType.ERROR, "Error in read: " + e.getMessage());
            throw e;
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public void write(byte[] bytes) throws TransportException {
        if (bytes == null || bytes.length == 0) {
            return;
        }

        writeLock.lock();
        try {
            ensureOpen();

            ByteBuffer writeBuffer = ByteBuffer.wrap(bytes);

            // Blocking write parks the virtual thread until the kernel send buffer accepts the
            // bytes — natural backpressure, no OP_WRITE registration or sleep loop needed.
            // (write() never returns -1; a broken/closed connection surfaces as
            // IOException/AsynchronousCloseException, both handled below.)
            while (writeBuffer.hasRemaining()) {
                socketChannel.write(writeBuffer);
            }

            LOGGER.trace("Wrote {} bytes", bytes.length);

            // Log the bytes to the audit log
            if (getAuditLog().isEnabled()) {
                getAuditLog().write(AuditLogEventType.OUTGOING_BYTES, "Write: " + StaticHelper.ENCODE_HEX(bytes));
            }
        } catch (AsynchronousCloseException e) {
            // A concurrent close() closed the channel while we were parked in write(): normal shutdown.
            if (!open.get()) {
                return;
            }
            getAuditLog().write(AuditLogEventType.ERROR, "Error in write: " + e.getMessage());
            throw new TransportException("Failed to write data", e);
        } catch (IOException e) {
            getAuditLog().write(AuditLogEventType.ERROR, "Error in write: " + e.getMessage());
            throw new TransportException("Failed to write data", e);
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public void close() throws TransportException {
        // CAS so concurrent/repeated close() calls run the shutdown exactly once.
        if (!open.compareAndSet(true, false)) {
            return;
        }

        // Intentionally takes NO locks: closing the channel is what unblocks a parked read()/write().
        // Acquiring writeLock first would deadlock against a writer parked in a blocking write().
        try {
            socketChannel.close();
            // Only log/audit a successful close here; on failure the catch reports ERROR
            // and rethrows, so emitting CLOSE in finally would falsely signal a clean close.
            LOGGER.debug("TCP connection closed");
            getAuditLog().write(AuditLogEventType.CLOSE, "Closed");
        } catch (IOException e) {
            getAuditLog().write(AuditLogEventType.ERROR, "Error in close: " + e.getMessage());
            throw new TransportException("Failed to close connection", e);
        } finally {
            // Always join the read loop, regardless of whether the channel closed cleanly.
            // Skip the self-join when close() runs on the read thread itself (e.g. a
            // disconnect/data listener calls close()) — joining yourself only stalls for the
            // timeout and the loop already exits once open is false.
            if (readThread != null && Thread.currentThread() != readThread) {
                try {
                    readThread.join(1000);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }
    }

    /**
     * Ensures the connection is still open, throws exception otherwise.
     */
    private void ensureOpen() throws TransportException {
        if (!isOpen()) {
            throw new TransportException("Transport is closed");
        }
    }

    // ========== AsyncTransportInstance Implementation ==========

    @Override
    public void registerDataListener(Runnable listener) {
        this.dataListener = listener;
        LOGGER.debug("Data listener registered");
    }

    @Override
    public void removeDataListener() {
        this.dataListener = null;
        LOGGER.debug("Data listener removed");
    }

    @Override
    public void registerDisconnectListener(Consumer<Throwable> listener) {
        this.disconnectListener = listener;
        LOGGER.debug("Disconnect listener registered");
    }

    @Override
    public void removeDisconnectListener() {
        this.disconnectListener = null;
        LOGGER.debug("Disconnect listener removed");
    }

    /**
     * Notifies the disconnect listener if one is registered.
     *
     * @param cause the exception that caused the disconnect, or null for graceful close
     */
    private void notifyDisconnect(Throwable cause) {
        Consumer<Throwable> listener = disconnectListener;
        if (listener != null) {
            try {
                listener.accept(cause);
            } catch (Exception e) {
                LOGGER.error("Error in disconnect listener", e);
            }
        }
    }

    /**
     * Runs the listener on the read thread, guarding against a misbehaving listener so a thrown
     * exception cannot silently kill the read loop (mirrors {@link #notifyDisconnect}'s posture).
     */
    private void safeRun(Runnable listener) {
        if (listener == null) {
            return;
        }
        try {
            listener.run();
        } catch (Throwable t) {
            LOGGER.error("Data listener failed", t);
        }
    }

    /**
     * Per-connection read loop on a virtual thread: blocking read into the ring buffer, then
     * notify the data listener. No selector, no polling.
     */
    private void runReadLoop() {
        LOGGER.debug("Read loop started");
        try {
            while (open.get()) {
                int free = ringBuffer.remainingForWriting();
                if (free == 0) {
                    // Backpressure: the ring buffer is full and the consumer has not drained yet.
                    // Park briefly and re-check; never disconnect (only the codec knows frame
                    // boundaries, and cross-thread consumers like COTP can legitimately lag).
                    LockSupport.parkNanos(200_000L);
                    continue;
                }

                // Bound the read to free ring-buffer space so the buffer can never overflow.
                readBuffer.clear();
                readBuffer.limit(Math.min(readBuffer.capacity(), free));

                int bytesRead = socketChannel.read(readBuffer);  // parks vthread; releases carrier (JDK21)
                if (bytesRead == -1) {
                    // Connection closed gracefully by remote
                    LOGGER.info("Connection closed by remote");
                    open.set(false);
                    notifyDisconnect(null);
                    break;
                }
                if (bytesRead == 0) {
                    // A blocking read effectively never returns 0; harmless guard.
                    continue;
                }

                readBuffer.flip();
                readLock.lock();
                try {
                    ringBuffer.write(readBuffer);
                } finally {
                    readLock.unlock();
                }

                // Notify OUTSIDE readLock — the data is already in the ring buffer.
                safeRun(dataListener);
            }
        } catch (IOException e) {
            // If open==false, an intentional close() closed the channel (AsynchronousCloseException) —
            // a normal shutdown, not a disconnect. Only a failure while still open is a real disconnect.
            if (open.get()) {
                getAuditLog().write(AuditLogEventType.ERROR, "Error in runReadLoop: " + e.getMessage());
                LOGGER.error("Error in read loop", e);
                open.set(false);
                notifyDisconnect(e);
            }
        }
        LOGGER.debug("Read loop stopped");
    }

}

File: plc4net/transports/tcp/TcpTransportInstance.cs
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//

using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using org.apache.plc4net.spi.transports;

namespace org.apache.plc4net.transports.tcp
{
    /// <summary>
    /// TCP transport: one background read loop per connection feeding a ring buffer.
    /// </summary>
    /// <remarks>
    /// The Java SPI3 transport uses a virtual thread per connection blocked in a socket
    /// read. .NET has no virtual threads, but async I/O gives the same result: the read
    /// loop awaits <c>ReceiveAsync</c> and holds no thread while idle. The observable
    /// contract is the same — the loop fills the ring buffer and then invokes the data
    /// listener.
    /// <para>
    /// The transport SPI is synchronous: opening a connection returns a connected
    /// instance, so the constructor has to block its caller until the connect is done.
    /// It does that in exactly one place, <see cref="Connect"/>. The connect itself is
    /// async I/O with a cancellation token that fires when the connect timeout expires,
    /// so an attempt that runs out of time is actually aborted rather than abandoned.
    /// The work runs on the thread pool, so no synchronization context on the calling
    /// thread can ever be needed to complete it.
    /// </para>
    /// <para>
    /// <see cref="Close"/> never waits for the read loop. Closing the socket is what
    /// ends the loop, and it finishes on its own thread.
    /// </para>
    /// </remarks>
    public class TcpTransportInstance : BaseTransportInstance, IAsyncTransportInstance
    {
        private readonly Socket _socket;
        private readonly RingBuffer _ringBuffer;
        private readonly object _readLock = new object();
        private readonly object _writeLock = new object();
        private readonly CancellationTokenSource _shutdown = new CancellationTokenSource();
        private readonly Task _readLoop;

        // Whether the connection is usable. Cleared by Close() and by the read loop
        // when the peer goes away. Deliberately separate from _released: the peer can
        // end the connection without anyone having released our end of it yet.
        private int _open = 1;

        // One-shot guard for ReleaseResources(). Every path out of the connection ends
        // there, and only the first one through does the work.
        private int _released;

        // Set by Close() before it touches the socket: a write that fails while this
        // is 1 is failing during our own shutdown, so the write path treats it as a
        // normal close rather than an error, whether the failure came from the
        // shutdown itself or from a peer that happened to go away at the same time.
        private int _closing;

        private volatile Action? _dataListener;
        private volatile Action<Exception?>? _disconnectListener;

        public TcpTransportInstance(IPEndPoint remoteAddress, TcpTransportConfiguration configuration)
            : base(configuration)
        {
            if (remoteAddress == null)
            {
                throw new ArgumentNullException(nameof(remoteAddress));
            }

            _ringBuffer = new RingBuffer(configuration.ReceiveBufferSize);
            _socket = Connect(remoteAddress, configuration);
            RemoteAddress = remoteAddress;
            LocalAddress = _socket.LocalEndPoint as IPEndPoint;

            // Started last so a throw during setup cannot leak a running loop.
            _readLoop = Task.Run(RunReadLoopAsync);
        }

        public IPEndPoint RemoteAddress { get; }

        public IPEndPoint? LocalAddress { get; }

        /// <summary>
        /// Opens the connection and waits for it, but no longer than the configured
        /// connect timeout.
        /// </summary>
        private static Socket Connect(IPEndPoint remoteAddress, TcpTransportConfiguration configuration)
        {
            CancellationTokenSource? deadline = null;
            try
            {
                deadline = new CancellationTokenSource(configuration.ConnectTimeoutMillis);
                var token = deadline.Token;
                return Task.Run(() => ConnectAsync(remoteAddress, configuration, token), token)
                    .GetAwaiter().GetResult();
            }
            catch (OperationCanceledException) when (deadline?.IsCancellationRequested == true)
            {
                throw new TransportException(
                    $"Failed to connect to {remoteAddress.Address}:{remoteAddress.Port} - " +
                    $"timed out after {configuration.ConnectTimeout} ms.");
            }
            catch (Exception e) when (!(e is TransportException))
            {
                throw new TransportException(
                    $"Failed to connect to {remoteAddress.Address}:{remoteAddress.Port} - {e.Message}", e);
            }
            finally
            {
                deadline?.Dispose();
            }
        }

        /// <summary>
        /// Creates a socket, applies the configured options and connects it. Cancelling
        /// <paramref name="cancellationToken"/> aborts the connect and disposes the socket.
        /// </summary>
        internal static async Task<Socket> ConnectAsync(
            IPEndPoint remoteAddress, TcpTransportConfiguration configuration, CancellationToken cancellationToken)
        {
            var socket = new Socket(remoteAddress.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
            try
            {
                if (!string.IsNullOrEmpty(configuration.LocalAddress))
                {
                    socket.Bind(new IPEndPoint(IPAddress.Parse(configuration.LocalAddress), configuration.LocalPort));
                }

                socket.NoDelay = configuration.TcpNoDelay;
                socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, configuration.KeepAlive);
                if (configuration.SendBufferSize > 0)
                {
                    socket.SendBufferSize = configuration.SendBufferSize;
                }
                if (configuration.ReceiveBufferSize > 0)
                {
                    socket.ReceiveBufferSize = configuration.ReceiveBufferSize;
                }

                await socket.ConnectAsync(remoteAddress, cancellationToken).ConfigureAwait(false);
                return socket;
            }
            catch
            {
                socket.Dispose();
                throw;
            }
        }

        public override bool IsOpen => Volatile.Read(ref _open) == 1 && _socket.Connected;

        /// <summary>
        /// Whether the socket has been shut down and disposed. Exposed so tests can tell
        /// "the peer went away" apart from "our end was actually released" - the two used
        /// to be the same flag, which is how the release came to be skipped.
        /// </summary>
        internal bool ResourcesReleased => Volatile.Read(ref _released) == 1;

        /// <summary>
        /// Completes when the read loop has finished. Exposed so tests can wait for the
        /// end of a shutdown that <see cref="Close"/> deliberately does not wait for.
        /// </summary>
        internal Task ReadLoopCompleted => _readLoop;

        /// <summary>
        /// Optional diagnostic output. When set, raw bytes written and read
        /// are logged as annotated hex.
        /// </summary>
        public TextWriter? DiagnosticOutput { get; set; }

        public override int GetNumBytesAvailable()
        {
            lock (_readLock)
            {
                return IsOpen || _ringBuffer.AvailableForReading > 0
                    ? _ringBuffer.AvailableForReading
                    : 0;
            }
        }

        public override byte[] PeekReadableBytes(int numBytes)
        {
            if (numBytes <= 0)
            {
                return Array.Empty<byte>();
            }
            lock (_readLock)
            {
                if (_ringBuffer.AvailableForReading < numBytes)
                {
                    throw new TransportException(
                        $"Requested {numBytes} bytes but only {_ringBuffer.AvailableForReading} available");
                }
                return _ringBuffer.Peek(numBytes);
            }
        }

        public override byte[] Read(int numBytes)
        {
            if (numBytes <= 0)
            {
                return Array.Empty<byte>();
            }
            byte[] result;
            lock (_readLock)
            {
                if (_ringBuffer.AvailableForReading < numBytes)
                {
                    throw new TransportException(
                        $"Requested {numBytes} bytes but only {_ringBuffer.AvailableForReading} available");
                }
                result = _ringBuffer.Read(numBytes);
            }
            WriteHex("TCP RECV", result);
            return result;
        }

        public override void Write(byte[] bytes)
        {
            if (bytes == null || bytes.Length == 0)
            {
                return;
            }

            lock (_writeLock)
            {
                EnsureOpen();
                try
                {
                    var offset = 0;
                    while (offset < bytes.Length)
                    {
                        var sent = _socket.Send(bytes, offset, bytes.Length - offset, SocketFlags.None);
                        if (sent == 0)
                        {
                            // A blocking Send only returns 0 once the connection is gone.
                            // Looping again would spin forever, since offset cannot move.
                            throw new TransportException(
                                $"Connection closed after writing {offset} of {bytes.Length} bytes.");
                        }
                        offset += sent;
                    }
                    WriteHex("TCP SEND", bytes);
                }
                catch (ObjectDisposedException) when (Volatile.Read(ref _open) == 0)
                {
                    // A concurrent Close() disposed the socket mid-write: normal shutdown.
                }
                catch (SocketException) when (Volatile.Read(ref _closing) == 1)
                {
                    // Close() shuts the socket down before disposing it, so a write
                    // blocked on a full send buffer wakes up with SocketException
                    // (an abort/interrupt style error, or a broken pipe) rather than
                    // ObjectDisposedException. _closing is set before the shutdown
                    // starts, so seeing it 1 means our own close caused this: normal
                    // shutdown, same as the disposed case above. A peer failure that
                    // races a concurrent close gets reported the same way; a peer
                    // reset without any local close still surfaces below.
                }
                catch (SocketException e)
                {
                    throw new TransportException("Failed to write data", e);
                }
            }
        }

        public override void Close()
        {
            Interlocked.Exchange(ref _closing, 1);
            Volatile.Write(ref _open, 0);
            ReleaseResources();
        }

        /// <summary>
        /// Shuts the socket down and disposes it, exactly once.
        /// </summary>
        /// <remarks>
        /// Reached from three directions: an explicit <see cref="Close"/>, a
        /// <see cref="BaseTransportInstance.Dispose()"/>, or the read loop finishing —
        /// which happens on its own whenever the peer hangs up. Keeping the guard here
        /// rather than on the open flag is what makes the last of those release
        /// anything: the loop clears the open flag as it leaves, so a Close() arriving
        /// afterwards would otherwise see an already-closed connection and return
        /// without ever disposing the socket.
        /// <para>
        /// Takes no locks and does not wait for the read loop. Closing the socket is what
        /// unblocks the loop, and taking <c>_writeLock</c> first would deadlock against a
        /// blocked writer. The loop ends on its own thread; a listener callback that is
        /// still running when this returns simply finishes first.
        /// </para>
        /// </remarks>
        private void ReleaseResources()
        {
            if (Interlocked.CompareExchange(ref _released, 1, 0) != 0)
            {
                return;
            }

            _shutdown.Cancel();
            try
            {
                _socket.Shutdown(SocketShutdown.Both);
            }
            catch (SocketException)
            {
                // Already torn down by the peer; nothing to do.
            }
            catch (ObjectDisposedException)
            {
            }
            finally
            {
                _socket.Dispose();
                // The read loop may still be reading the token. It treats the resulting
                // ObjectDisposedException as the normal end of a shutdown.
                _shutdown.Dispose();
            }
        }

        public void RegisterDataListener(Action listener) => _dataListener = listener;

        public void RemoveDataListener() => _dataListener = null;

        public void RegisterDisconnectListener(Action<Exception?> listener) => _disconnectListener = listener;

        public void RemoveDisconnectListener() => _disconnectListener = null;

        private void NotifyDisconnect(Exception? cause)
        {
            var listener = _disconnectListener;
            if (listener == null)
            {
                return;
            }
            try
            {
                listener(cause);
            }
            catch
            {
                // A misbehaving listener must not take down the read loop.
            }
        }

        private void NotifyData()
        {
            var listener = _dataListener;
            if (listener == null)
            {
                return;
            }
            try
            {
                listener();
            }
            catch
            {
                // Same reasoning as NotifyDisconnect.
            }
        }

        private async Task RunReadLoopAsync()
        {
            var buffer = new byte[8192];
            try
            {
                while (Volatile.Read(ref _open) == 1)
                {
                    int free;
                    lock (_readLock)
                    {
                        free = _ringBuffer.RemainingForWriting;
                    }
                    if (free == 0)
                    {
                        // Backpressure: the codec has not drained yet. Wait and retry
                        // rather than disconnect — only the codec knows frame boundaries.
                        await Task.Delay(1, _shutdown.Token).ConfigureAwait(false);
                        continue;
                    }

                    var toRead = Math.Min(buffer.Length, free);
                    var bytesRead = await _socket
                        .ReceiveAsync(buffer.AsMemory(0, toRead), SocketFlags.None, _shutdown.Token)
                        .ConfigureAwait(false);

                    if (bytesRead == 0)
                    {
                        // Orderly shutdown by the remote end. When our own Close() got
                        // here first, the zero is just its Shutdown coming back and
                        // there is nothing to report.
                        if (Interlocked.Exchange(ref _open, 0) == 1)
                        {
                            NotifyDisconnect(null);
                        }
                        return;
                    }

                    lock (_readLock)
                    {
                        _ringBuffer.Write(buffer, 0, bytesRead);
                    }

                    // Notify outside the lock; the bytes are already buffered.
                    NotifyData();
                }
            }
            catch (OperationCanceledException)
            {
                // Close() cancelled us: normal shutdown.
            }
            catch (ObjectDisposedException)
            {
                // Close() disposed the socket or the token while we were using them.
            }
            catch (Exception e)
            {
                // Only a failure while still open counts as a disconnect.
                if (Interlocked.Exchange(ref _open, 0) == 1)
                {
                    NotifyDisconnect(e);
                }
            }
            finally
            {
                // However this loop ends, our end of the connection is finished with.
                // When Close() started the teardown this is a no-op; when the peer did,
                // this is the only thing that releases the socket. Clear _open so that
                // IsOpen does not read _socket.Connected on a disposed socket, even if a
                // future catch block forgets to.
                Volatile.Write(ref _open, 0);
                ReleaseResources();
            }
        }

        private void WriteHex(string label, byte[] bytes)
        {
            var diag = DiagnosticOutput;
            if (diag == null) return;
            diag.WriteLine($"┌ {label} ({bytes.Length} bytes)");
            for (int i = 0; i < bytes.Length; i += 16)
            {
                diag.Write("│ ");
                for (int j = i; j < Math.Min(i + 16, bytes.Length); j++)
                    diag.Write($"{bytes[j]:X2} ");
                diag.WriteLine();
            }
            diag.WriteLine("└");
            diag.Flush();
        }
    }
}